lib/_tls_wrap.js

```diff
@@ -37,211 +37,243 @@ const tls_wrap = process.binding('tls_wrap');
+
+const noop = () => {};
 
 function onhandshakestart() {
   debug('onhandshakestart');
 
-  var self = this;
-  var ssl = self._handle;
-  var now = Timer.now();
+  const owner = this.owner;
+  const now = Timer.now();
 
-  assert(now >= ssl.lastHandshakeTime);
+  assert(now >= this.lastHandshakeTime);
 
-  if ((now - ssl.lastHandshakeTime) >= tls.CLIENT_RENEG_WINDOW * 1000) {
-    ssl.handshakes = 0;
+  if ((now - this.lastHandshakeTime) >= tls.CLIENT_RENEG_WINDOW * 1000) {
+    this.handshakes = 0;
   }
 
-  var first = (ssl.lastHandshakeTime === 0);
-  ssl.lastHandshakeTime = now;
+  const first = (this.lastHandshakeTime === 0);
+  this.lastHandshakeTime = now;
   if (first) return;
 
-  if (++ssl.handshakes > tls.CLIENT_RENEG_LIMIT) {
+  if (++this.handshakes > tls.CLIENT_RENEG_LIMIT) {
     // Defer the error event to the next tick. We're being called from OpenSSL's
     // state machine and OpenSSL is not re-entrant. We cannot allow the user's
     // callback to destroy the connection right now, it would crash and burn.
-    setImmediate(function() {
-      var err = new errors.Error('ERR_TLS_SESSION_ATTACK');
-      self._emitTLSError(err);
-    });
+    setImmediate(emitSessionAttackError, owner);
   }
 
-  if (this[kDisableRenegotiation] && ssl.handshakes > 0) {
+  if (owner[kDisableRenegotiation] && this.handshakes > 0) {
     const err = new Error('TLS session renegotiation disabled for this socket');
-    self._emitTLSError(err);
+    owner._emitTLSError(err);
   }
 }
 
+function emitSessionAttackError(self) {
+  self._emitTLSError(new errors.Error('ERR_TLS_SESSION_ATTACK'));
+}
 
 function onhandshakedone() {
-  // for future use
   debug('onhandshakedone');
-  this._finishInit();
+
+  const owner = this.owner;
+
+  // `newSession` callback wasn't called yet
+  if (owner._newSessionPending) {
+    owner._securePending = true;
+    return;
+  }
+
+  owner._finishInit();
 }
 
 
-function loadSession(self, hello, cb) {
+function loadSession(hello) {
+  const owner = this.owner;
+
   var once = false;
   function onSession(err, session) {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return loadSessionDone(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return loadSessionDone(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return loadSessionDone(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
-    self._handle.loadSession(session);
-    cb(null);
+    owner._handle.loadSession(session);
+    loadSessionDone(null, owner);
   }
 
   if (hello.sessionId.length <= 0 ||
       hello.tlsTicket ||
-      self.server &&
-      !self.server.emit('resumeSession', hello.sessionId, onSession)) {
-    cb(null);
+      owner.server &&
+      !owner.server.emit('resumeSession', hello.sessionId, onSession)) {
+    loadSessionDone(null, owner);
```

Comment:
All entries of loadSessionDone have a a single line that they would call in the function. This can always be inlined directly. There is no need for calling that function.